diff --git a/tests/run/cpp_inherit_overload_T800.pyx b/tests/run/cpp_inherit_overload_T800.pyx
new file mode 100644
index 0000000..e89163e
--- /dev/null
+++ b/tests/run/cpp_inherit_overload_T800.pyx
@@ -0,0 +1,32 @@
+# mode: run
+# tag: cpp
+
+cdef extern from "cpp_inherit_overload_T800_helper.h":
+    cdef cppclass Base:
+        int foo(double)
+        int foo(double*)
+
+    cdef cppclass Derived(Base):
+        pass
+
+def test_base():
+    """
+    >>> test_base()
+    1
+    2
+    """
+    cdef Base obj
+    cdef double dummy = 1.
+    print obj.foo(dummy)
+    print obj.foo(&dummy)
+
+def test_derived():
+    """
+    >>> test_derived()
+    1
+    2
+    """
+    cdef Derived obj
+    cdef double dummy = 1.
+    print obj.foo(dummy)
+    print obj.foo(&dummy)
diff --git a/tests/run/cpp_inherit_overload_T800_helper.h b/tests/run/cpp_inherit_overload_T800_helper.h
new file mode 100644
index 0000000..3004586
--- /dev/null
+++ b/tests/run/cpp_inherit_overload_T800_helper.h
@@ -0,0 +1,15 @@
+class Base
+{
+	public:
+		int foo(double) {
+			return 1;
+		}
+
+		int foo(double*) {
+			return 2;
+		}
+};
+
+class Derived : public Base
+{
+};
