diff -r b577bd2ec4ba Cython/Compiler/ExprNodes.py
--- a/Cython/Compiler/ExprNodes.py	Wed Dec 17 01:43:58 2008 -0800
+++ b/Cython/Compiler/ExprNodes.py	Mon Jan 05 22:59:58 2009 +0200
@@ -4904,7 +4904,7 @@
     else if (Py_TYPE(o)->tp_as_sequence && Py_TYPE(o)->tp_as_sequence->sq_item && (likely(i >= 0) || !is_unsigned))
         r = PySequence_GetItem(o, i);
     else {
-        PyObject *j = (likely(i >= 0) || !is_unsigned) ? PyInt_FromLong(i) : PyLong_FromUnsignedLongLong((sizeof(unsigned long long) > sizeof(Py_ssize_t) ? (1ULL << (sizeof(Py_ssize_t)*8)) : 0) + i);
+        PyObject *j = (likely(i >= 0) || !is_unsigned) ? PyInt_FromLong(i) : PyLong_FromUnsignedLongLong((sizeof(unsigned PY_LONG_LONG) > sizeof(Py_ssize_t) ? (((unsigned PY_LONG_LONG) 1) << (sizeof(Py_ssize_t)*8)) : 0) + i);
         if (!j)
             return 0;
         r = PyObject_GetItem(o, j);
@@ -4931,7 +4931,7 @@
     else if (Py_TYPE(o)->tp_as_sequence && Py_TYPE(o)->tp_as_sequence->sq_ass_item && (likely(i >= 0) || !is_unsigned))
         r = PySequence_SetItem(o, i, v);
     else {
-        PyObject *j = (likely(i >= 0) || !is_unsigned) ? PyInt_FromLong(i) : PyLong_FromUnsignedLongLong((sizeof(unsigned long long) > sizeof(Py_ssize_t) ? (1ULL << (sizeof(Py_ssize_t)*8)) : 0) + i);
+        PyObject *j = (likely(i >= 0) || !is_unsigned) ? PyInt_FromLong(i) : PyLong_FromUnsignedLongLong((sizeof(unsigned PY_LONG_LONG) > sizeof(Py_ssize_t) ? (((unsigned PY_LONG_LONG) 1) << (sizeof(Py_ssize_t)*8)) : 0) + i);
         if (!j)
             return -1;
         r = PyObject_SetItem(o, j, v);
@@ -4952,7 +4952,7 @@
     if (Py_TYPE(o)->tp_as_sequence && Py_TYPE(o)->tp_as_sequence->sq_ass_item && (likely(i >= 0) || !is_unsigned))
         r = PySequence_DelItem(o, i);
     else {
-        PyObject *j = (likely(i >= 0) || !is_unsigned) ? PyInt_FromLong(i) : PyLong_FromUnsignedLongLong((sizeof(unsigned long long) > sizeof(Py_ssize_t) ? (1ULL << (sizeof(Py_ssize_t)*8)) : 0) + i);
+        PyObject *j = (likely(i >= 0) || !is_unsigned) ? PyInt_FromLong(i) : PyLong_FromUnsignedLongLong((sizeof(unsigned PY_LONG_LONG) > sizeof(Py_ssize_t) ? (((unsigned PY_LONG_LONG) 1) << (sizeof(Py_ssize_t)*8)) : 0) + i);
         if (!j)
             return -1;
         r = PyObject_DelItem(o, j);
