diff -r 5a7f83a639f7 Cython/Compiler/ParseTreeTransforms.py
--- a/Cython/Compiler/ParseTreeTransforms.py	Tue Oct 14 09:08:43 2008 +0200
+++ b/Cython/Compiler/ParseTreeTransforms.py	Wed Oct 15 20:10:13 2008 +0200
@@ -183,9 +183,10 @@ class PostParse(CythonTransform):
                                     handler(decl)
                                     continue # Remove declaration
                             raise PostParseError(decl.pos, ERR_CDEF_INCLASS)
+                        first_assignment = self.scope_type != 'module'
                         stats.append(SingleAssignmentNode(node.pos,
                             lhs=NameNode(node.pos, name=declbase.name),
-                            rhs=declbase.default, first=True))
+                            rhs=declbase.default, first=first_assignment))
                         declbase.default = None
                 newdecls.append(decl)
             node.declarators = newdecls
