diff -r 57467649c34c Cython/Compiler/Optimize.py
--- a/Cython/Compiler/Optimize.py	Wed Oct 15 12:10:09 2008 -0300
+++ b/Cython/Compiler/Optimize.py	Wed Oct 15 13:05:13 2008 -0300
@@ -154,9 +154,12 @@ class FinalOptimizePhase(Visitor.CythonT
             lhs = node.lhs
             lhs.lhs_of_first_assignment = True
             if isinstance(lhs, ExprNodes.NameNode) and lhs.entry.type.is_pyobject:
-                # Have variable initialized to 0 rather than None
-                lhs.entry.init_to_none = False
                 lhs.entry.init = 0
+                if lhs.entry.scope.is_module_scope:
+                    lhs.entry.init_to_none = True
+                    lhs.lhs_of_first_assignment = False
+                else:
+                    lhs.entry.init_to_none = False
         return node
 
     def visit_SimpleCallNode(self, node):
